qiskit_experiments/framework/cache_method.py

```diff
@@ -0,0 +1,144 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Method decorator for caching regular methods in class instances.
+"""
+
+from typing import Union, Dict, Callable
+import functools
+
+
+def cache_method(
+    cache: Union[Dict, str] = "_cache", cache_args: bool = True, require_hashable: bool = True
+) -> Callable:
+    """Decorator for caching class instance methods.
+
+    Args:
+        cache: The cache or cache attribute name to use. If a dict it will
```

Comment:
Seems like the definition is little vague because it says we can use any object (usually a dict) but the requirement for the object is not written here. If this must be a dict we should type check otherwise we must describe the requirement, e.g. subclass of `MutableMapping` , or some class that implements the xxx and yyy method...

Reply:
The type hint and doc string both say dict or string, so I don't follow how that is vague?

Technically you are correct that it doesn't explicitly check for a dict, so it could be something more general that behaves like a dict (technically it just needs a get/set item method that works with hashable objects)

Reply:
I think just saying dictionary or string should be enough. If a developer want to know the possibility of non-dictionary object, they must read the code.